### PR TITLE
adds error handling for grpc errors

### PR DESCRIPTION
**Pull request by mchestnut91**
This PR is a follow up from [this ticket](https://app.clubhouse.io/pingthings-ws/story/4093/enumerate-btrdb-grpc-error-codes) and is meant to verify that btrdb-server error codes are properly sent through the python bindings and displayed to the user. In addition, this PR attempts to catch any grpc errors and present them in a way that more user friendly.

I found that the bindings correctly display the errors defined in btrdb-server [here](https://github.com/PingThingsIO/btrdb-server/blob/87391ca15128c39a25f79262b0d8b61e52d319ec/bte/errors.go). Here are a few examples:

I tried to create a stream with a uuid that already exists:
```
btrdb.exceptions.BTrDBError: [417] a stream already exists with that uuid
```

I tried to update a stream’s metadata with an invalid tag:
```
btrdb.exceptions.BTrDBError: [408] (408: tag key "chickens" is invalid)
```

I tested a few more of these, in addition to the few that David tested, and they all seemed to work as expected, and IMO these errors are nice looking and easy to understand. All of these errors are generated by checking the result status after a grpc request is completed with a line that looks like this: `BTrDBError.checkProtoStat(result.stat)`

This all works as expected, but there are some cases where it encounters an error during the actual grpc request. When this happens, the program panics and returns an ugly grpc error like this:
```_InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
    status = StatusCode.UNKNOWN
    details = "[404] stream does not exist"
    debug_error_string = "{"created":"@1612962972.818625965","description":"Error received from peer ipv4:10.30.10.12:4411","file":"src/core/lib/surface/call.cc","file_line":1062,"grpc_message":"[404] stream does not exist","grpc_status":2}"
```

To fix this I put a try/except around each grpc call in `endpoint.py` that checks for any `grpc.RpcError`. I also wrote an error handling function that checks the details field for any known errors and returns a prettier error. Currently the two errors that I’ve identified are: 
1. the stream does not exist 
2. there is an issue with the user’s internet/VPN connection. 

Here is what the previous error looks like with these changes:

```btrdb.exceptions.StreamNotFoundError: Stream not found with provided uuid```

I imagine that there are other things that would cause grpc errors, but I cannot think of them right now. I’ve enlisted the help of the DS team to create a ticket whenever they encounter a grpc error so I can go back and trap it and return something prettier.

### CODE REVIEW

File: btrdb/exceptions.py [hunk 1]
```diff
@@ -11,11 +11,22 @@
 Module for custom exceptions
 """
 
+# Handles grpc errors
+def handle_grpc_error(err):
+    if err.details() == "[404] stream does not exist":
```
Comment (oridb) [inline btrdb/exceptions.py:16]: Can we use the error codes here?

https://github.com/PingThingsIO/btrdb-server/blob/9d57e4c97380ebf5160b0752becead041efbe86b/bte/errors.go

File: btrdb/endpoint.py [hunk 2]
```diff
@@ -25,45 +25,80 @@
...
-    def rawValues(self, uu, start, end, version = 0):
-        params = btrdb_pb2.RawValuesParams(uuid = uu.bytes, start = start, end = end, versionMajor = version)
-        for result in self.stub.RawValues(params):
-            BTrDBError.checkProtoStat(result.stat)
-            yield result.values, result.versionMajor
+    def rawValues(self, uu, start, end, version=0):
+        params = btrdb_pb2.RawValuesParams(
+            uuid=uu.bytes, start=start, end=end, versionMajor=version
+        )
+        try:
```
Comment (oridb) [inline btrdb/endpoint.py:45]: It seems like the try/catch boilerplate could be a decorator:
```
@error_wrap
def ...
```
where we do something like:
```
def error_wrap(fn):
      def wrap(*args, **kwargs):
           try: fn(*args, **kwargs)
           except grpc.Error as e: handle()
      return wrap

**Comment by looselycoupled** [comment]
Great work, but the overall goal is to make new errors like `btrdb.exceptions.StreamNotFoundError` so that the users can easily capture such errors.  We do not want to just reraise the GRPC error - we want to hide the GRPCness of the bindings.

**Comment by mchestnut91** [comment]
@looselycoupled maybe we can chat about what you have in mind, because that's what I thought I was doing in my `handle_grpc_errors()` function. It seems to me like we are properly handling errors as long as they are able to complete the grpc call. I have identified two cases where the code fails during the actual grpc request, and we can trap those errors and return either `btrdb.exceptions.StreamNotFoundError` or `btrdb.exceptions.ConnectionError`. In these cases the user would not see the grpc errors

File: btrdb/endpoint.py (hunk 2 above)
Comment (mchestnut91) [reply btrdb/endpoint.py:45]: Yea - this is a good idea. I'll work on this

**Comment by looselycoupled** [comment]
@mchestnut91 cool - let's hop on a quick call Wednesday!

**Comment by mchestnut91** [comment]
Updates to this PR include:
* Created custom BTrDB exceptions for the errors that I deemed most likely to occur.
* Created a list of error codes that will raise a generic `BTrDBServerError`. Note that there are a few of the [btrdb-server error codes](https://github.com/PingThingsIO/btrdb-server/blob/v5/bte/errors.go) that I did not include because the server did not seem to use them. I created a [Notion page](https://www.notion.so/pingthings2/Btrdb-python-error-codes-09e02801ce5240ed802c31a876542e97) that lays out how I classified each error code. Definitely let me know if you think something is not categorized correctly
* Created decorators for handling grpc errors for the functions in `endpoint.py` per @oridb's suggestion
* Updated tests to work with new exceptions
* Replaced instances of general errors in `stream.py` with their custom BTrDB errors. e.g. replaced `TypeError` with `BTRDBTypeError`

As a possible next step, I've made myself a ticket to dig into the btrdb-server code to see if we can figure out why it sends some errors with `StatusCode.UNKNOWN`. We are currently handling this errors by string matching the details field of the error, but this is obviously not ideal and would be better to fix on the server side

**Comment by looselycoupled** [comment]
Just leaving a comment that I still need to review this.

File: btrdb/errors.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    ----------
+    stat: btrdb_pb2.Status
+    """
+    code = stat.code
+    if code != 0:
+        if code in BTRDB_ERRORS:
+            raise BTRDB_ERRORS[code](stat.msg)
+        elif code in BTRDB_SERVER_ERRORS:
+            raise BTRDBServerError(stat.msg)
+        raise BTrDBError(stat.msg)
```
Comment (looselycoupled) [inline btrdb/errors.py:159]: "Generally" the terms Errors and Exceptions are interchangeable in Python.  Let's move all this code into `exceptions.py`.

File: btrdb/exceptions.py [hunk 4]
```diff
@@ -7,81 +7,138 @@
...
+    """
+    An error has occurred while trying to establish a connection with BTrDB.
+    """
+    pass
 
-    def isError(self):
-        return self.code != 0
+class StreamNotFoundError(BTrDBError):
+    """
+    A problem interacting with the BTrDB server.
```
Comment (looselycoupled) [inline btrdb/exceptions.py:28]: Can we provide more descriptive help here?

File: btrdb/exceptions.py [hunk 5]
```diff
@@ -7,81 +7,138 @@
...
+
+class InvalidTimeRange(BTRDBValueError):
+    """
+    Insert contains a timestamp outside the range of (btrdb.MINIMUM_TIME, btrdb.MAXIMUM_TIME)
+    """
+    pass
+
+class InvalidPointWidth(BTRDBValueError):
+    """
+    Valid pointwidths are (0, 64)
```
Comment (looselycoupled) [inline btrdb/exceptions.py:112]: `Valid pointwidths are whole numbers between 0 and 64 (exclusive)`

Note, I forget if 64 ends up being valid.

File: btrdb/exceptions.py [hunk 6]
```diff
@@ -7,81 +7,138 @@
...
+    """
+    pass
+
+class BadValue(BTRDBValueError):
+    """
+    Returned when you try to insert None values
```
Comment (looselycoupled) [inline btrdb/exceptions.py:118]: Returned when you try to insert non-float values such as `None`.

File: btrdb/exceptions.py [hunk 7]
```diff
@@ -7,81 +7,138 @@
...
+
+class BadSQLValue(BTRDBValueError):
+    """
+    Invalid parameters have been passed to metadata db
+    """
+    pass
+
+class VersionNotAvailable(BTRDBValueError):
+    """
+    When querying a stream at a pruned version
```
Comment (looselycoupled) [inline btrdb/exceptions.py:136]: Can we standardize on the description?  Some start with "Returned if..." while other are like this.

Also, perhaps change to "... querying a stream at a pruned or otherwise invalid version number".

File: btrdb/errors.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+    """
+    # allows input func to keep its name and metadata
+    @wraps(fn)
+    def wrap(*args, **kwargs):
+        try:
+            if inspect.isgeneratorfunction(fn):
+                # putting yield directly in this function turns it into a generator,
+                # so keeping it separate
+                consume_generator(fn, *args, **kwargs)
+            return fn(*args, **kwargs)
```
Comment (looselycoupled) [inline btrdb/errors.py:110]: I'm confused at the return here as well as where consume_generator yields to...  Presumably consume_generator yields to wrap...  but wrap isnt iterating over the consume_generator.

Can you help me understand what's going on here?

File: btrdb/errors.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+    ----------
+    err: grpc.RpcError
+    """
+    details = err.details()
+    if details == "[404] stream does not exist":
+        raise StreamNotFoundError("Stream not found with provided uuid") from None
+    elif details == "failed to connect to all addresses":
+        raise ConnectionError("Failed to connect to BTrDB") from None
+    elif any(str(e) in err.details() for e in BTRDB_SERVER_ERRORS):
+        raise BTRDBServerError("An error has occured with btrdb-server") from None
```
Comment (looselycoupled) [inline btrdb/errors.py:138]: From None protects the call stack I am guessing?  Can you elaborate on this?

Comment (mchestnut91) [reply btrdb/errors.py:138]: Yea, I put this in to hide the grpc part of the error and only include the traceback from our code. Here is what it looks like when I take the `From None` out:

```
Traceback (most recent call last):
  File "/Users/michael.chestnut/projects/btrdb-python/btrdb/exceptions.py", line 26, in consume_generator
    yield from fn(*args, **kwargs)
  File "/Users/michael.chestnut/projects/btrdb-python/btrdb/endpoint.py", line 57, in alignedWindows
    for result in self.stub.AlignedWindows(params):
  File "/Users/michael.chestnut/projects/btrdb-python/venv/lib/python3.7/site-packages/grpc/_channel.py", line 416, in __next__
    return self._next()
  File "/Users/michael.chestnut/projects/btrdb-python/venv/lib/python3.7/site-packages/grpc/_channel.py", line 706, in _next
    raise self
grpc._channel._MultiThreadedRendezvous: <_MultiThreadedRendezvous of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "failed to connect to all addresses"
	debug_error_string = "{"created":"@1616086776.568522000","description":"Failed to pick subchannel","file":"src/core/ext/filters/client_channel/client_channel.cc","file_line":4134,"referenced_errors":[{"created":"@1616086776.568515000","description":"failed to connect to all addresses","file":"src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc","file_line":398,"grpc_status":14}]}"
>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test.py", line 26, in <module>
    vals, _ = zip(*stream.aligned_windows(btrdb.MINIMUM_TIME, btrdb.MAXIMUM_TIME, 30))
  File "/Users/michael.chestnut/projects/btrdb-python/btrdb/stream.py", line 674, in aligned_windows
    for stat_points, version in windows:
  File "/Users/michael.chestnut/projects/btrdb-python/btrdb/exceptions.py", line 28, in consume_generator
    yield handle_grpc_error(e)
  File "/Users/michael.chestnut/projects/btrdb-python/btrdb/exceptions.py", line 70, in handle_grpc_error
    raise ConnectionError("Failed to connect to BTrDB")
btrdb.exceptions.ConnectionError: Failed to connect to BTrDB
```
and with it we just get:
```
Traceback (most recent call last):
  File "test.py", line 26, in <module>
    vals, _ = zip(*stream.aligned_windows(btrdb.MINIMUM_TIME, btrdb.MAXIMUM_TIME, 30))
  File "/Users/michael.chestnut/projects/btrdb-python/btrdb/stream.py", line 674, in aligned_windows
    for stat_points, version in windows:
  File "/Users/michael.chestnut/projects/btrdb-python/btrdb/exceptions.py", line 28, in consume_generator
    yield handle_grpc_error(e)
  File "/Users/michael.chestnut/projects/btrdb-python/btrdb/exceptions.py", line 70, in handle_grpc_error
    raise ConnectionError("Failed to connect to BTrDB") from None
btrdb.exceptions.ConnectionError: Failed to connect to BTrDB
```

File: btrdb/errors.py (hunk 8 above)
Comment (mchestnut91) [reply btrdb/errors.py]: Apologies for the super long comment and this will likely include a lot of stuff you already know, but I wanted to outline my thought process as I was trying to make the error handling decorator work with the generator functions in `endpoint.py`.

I originally started with Ori's suggestion, which was something like this:

```python
def error_handler(fn):
    def wrap(*args, **kwargs):
        try:
            return fn(*args, **kwargs)
        except grpc.RpcError as e:
            handle_grpc_error(e)
    return wrap
```

This works well for non-generator functions, but when `fn` is a generator it gets returned
to the calling function, and when that function calls `next` and hits an error, there is nothing in place within the generator to catch the error, and the result is an unformatted grpc error. Not good.

So then I tried something like this:

```python
def error_handler(fn):
    @wraps(fn)
    def wrap(*args, **kwargs):
        if inspect.isgeneratorfunction(fn):
            try:
                yield from fn(*args, **kwargs)
            except grpc.RpcError as e:
                handle_grpc_error(e)
        else:
            try:
                return fn(*args, **kwargs)
            except grpc.RpcError as e:
                handle_grpc_error(e)
    return wrap
```
This worked when `fn` is a generator, but even when it isn't the presence of `yield` automatically made this function a generator, so the calling function would receive a generator instead of the returned value from `fn`. Also not good.

To get around this I moved the `yield from` part into a separate function. How this works is 
the generator gets returned to `wrap` and then gets returned to the calling function. Note that 
I was originally missing a return statement, so thank you for pointing that out. Here is what it looks like now:

```python
def consume_generator(fn, *args, **kwargs):
    # when a generator is passed back to the calling function, it may encounter an error
    # when trying to call next(), in that case we want to yield an Exception
    try:
        yield from fn(*args, **kwargs)
    except grpc.RpcError as e:
        handle_grpc_error(e)

def error_handler(fn):
    @wraps(fn)
    def wrap(*args, **kwargs):
        if inspect.isgeneratorfunction(fn):
            return consume_generator(fn, *args, **kwargs)
        try:
            return fn(*args, **kwargs)
        except grpc.RpcError as e:
            handle_grpc_error(e)
    return wrap
```

So now if I am doing an `aligned_windows()` query, the object that is returned in the call 

```python
windows = self._btrdb.ep.alignedWindows(self._uuid, start, end, pointwidth, version)
```

is  `<generator object consume_generator at 0x7fcb506e54d0>`. The calling function is then yielding from `consume_generator()` which is yielding from `fn`, and whenever it encounters an error it is able to handle with with `handle_grpc_error()`

tl;dr I had a lot of trouble getting the error handling decorator to work with the generator functions. I've done some testing and this seems to work, but it is admittedly a bit complicated. If you feel like this is too confusing, we could always go back to putting try/excepts back in the `endpoint.py` generator functions like I had before.

File: btrdb/exceptions.py (hunk 5 above)
Comment (mchestnut91) [reply btrdb/exceptions.py]: confirming that 64 is invalid

File: btrdb/exceptions.py [hunk 10]
```diff
@@ -7,81 +7,276 @@
...
 # ID: exceptions.py [] allen@pingthings.io $
 
-"""
-Module for custom exceptions
-"""
+##########################################################################
+## Imports
+##########################################################################
 
+import grpc
```
Comment (looselycoupled) [inline btrdb/exceptions.py:14]: If we're only using `grpc.RpcError` then let's change the import to `from grpc import RpcError`.

File: btrdb/exceptions.py [hunk 11]
```diff
@@ -7,81 +7,276 @@
...
     """
-    A requested profile could not be found in the credentials file.
+    Raised when asking for next/previous point and there isn't one.
     """
     pass
+
+
+##########################################################################
+## Exception mapping
+##########################################################################
```
Comment (looselycoupled) [inline btrdb/exceptions.py:235]: Let's move this to above the decorators section as constants should be defined early in the file.

File: btrdb/errors.py (hunk 9 above)
Comment (looselycoupled) [reply btrdb/errors.py]: Awesome, thanks!

File: btrdb/errors.py (hunk 8 above)
Comment (looselycoupled) [reply btrdb/errors.py]: Hmm..  I'm tempted to leave as you have it.  We would like the user called functions to get generators because we'd eventually like to have the bindings not fully materialize datasets before handing them on.  So provide a generator now and user code will not be impacted in a future enhancement.

**Comment by looselycoupled** [comment]
Gonna ask David and Ori to take another look but looking good.

**Review by davidkonigsberg — APPROVED** [review]
These errors are much more clear. I think this will really help with customers' understanding.

File: btrdb/exceptions.py (hunk 11 above)
Comment (mchestnut91) [reply btrdb/exceptions.py:235]: The problem with putting them before the exceptions are defined is I get errors for referencing the exceptions before assignment

Comment (looselycoupled) [reply btrdb/exceptions.py:235]: ahhh - yes I think I was only looking at the second part which just has the numbers.  Cool then.

**Review by oridb — APPROVED** [review]
